tools/reviewer/job/impl/SubmitterTask.java

```diff
@@ -0,0 +1,187 @@
+ * You may obtain a copy of the License at
+ *
+ *    https://www.apache.org/licenses/LICENSE-2.0
+ *
+ * Unless required by applicable law or agreed to in writing, software
+ * distributed under the License is distributed on an "AS IS" BASIS,
+ * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+ * See the License for the specific language governing permissions and
+ * limitations under the License.
+ */
+
+package com.google.startupos.tools.reviewer.job.impl;
+
+import com.google.common.flogger.FluentLogger;
+import com.google.startupos.common.FileUtils;
+import com.google.startupos.common.firestore.FirestoreClientFactory;
+import com.google.startupos.common.firestore.MessageWithId;
+import com.google.startupos.tools.reviewer.ReviewerProtos.CIRequest;
+import com.google.startupos.tools.reviewer.ReviewerProtos.CIResponse;
+import com.google.startupos.tools.reviewer.ReviewerProtos.Repo;
+import com.google.startupos.common.repo.GitRepo;
+import com.google.startupos.tools.reviewer.job.tasks.Task;
+
+import com.google.startupos.common.repo.GitRepoFactory;
+import com.google.startupos.tools.reviewer.localserver.service.Protos.Diff;
+import org.apache.commons.lang3.tuple.Pair;
+
+import javax.inject.Inject;
+import java.io.IOException;
+import java.util.ArrayList;
+import java.util.HashSet;
+import java.util.List;
+import java.util.concurrent.locks.ReentrantLock;
+
+public class SubmitterTask extends FirestoreTaskBase implements Task {
+  private static FluentLogger log = FluentLogger.forEnclosingClass();
+
+  private static String DIFF_PATH = "/reviewer/data/diff";
+  private static String CI_REQUESTS_PATH = "/reviewer/ci/requests";
+  private static String CI_RESPONSES_HISTORY_PATH = "/reviewer/ci/responses/%d/history";
+
+  private FileUtils fileUtils;
+  private GitRepoFactory gitRepoFactory;
+  private final ReentrantLock lock = new ReentrantLock();
+  private final HashSet<Long> ciSubmittedDiffIds = new HashSet<>();
+
+  @Inject
+  public SubmitterTask(
+      FileUtils fileUtils,
+      GitRepoFactory gitRepoFactory,
+      FirestoreClientFactory firestoreClientFactory) {
+    this.fileUtils = fileUtils;
+    this.gitRepoFactory = gitRepoFactory;
+    this.firestoreClientFactory = firestoreClientFactory;
+  }
+
+  @Override
+  public Boolean shouldRun() {
+    return !lock.isLocked();
+  }
+
+  @Override
+  public void run() {
+    if (lock.tryLock()) {
+      try {
+        initializeFirestoreClientIfNull();
+        // noinspection unchecked
+        List<Diff> diffs = (List) this.firestoreClient.listDocuments(DIFF_PATH, Diff.newBuilder());
+
+        if (!fileUtils.folderExists("submitter")) {
+          fileUtils.mkdirs("submitter");
+        }
+
+        diffs
+            .stream()
+            // get Diffs in SUBMITTING state
+            .filter(diff -> diff.getStatus().equals(Diff.Status.SUBMITTING))
+            // for each of them, get latest CIResponse (get or pop?)
```

Comment:
I think at this point, since we don't have a scheduler, let's get now, and remove after we succeed.

Reply:
@oferb I don't think CI responses should be removed at all. Shouldn't we track CI history for every diff?

Reply:
Ok, you're right.